lib/rspec/retry.rb

```diff
@@ -18,64 +18,114 @@ def self.apply
         # If no list of exceptions is provided and 'retry' > 1, we always retry.
         config.add_setting :exceptions_to_retry, :default => []
 
-        # context.example is deprecated, but RSpec.current_example is not
-        # available until RSpec 3.0.
-        fetch_current_example = RSpec.respond_to?(:current_example) ?
-          proc { RSpec.current_example } : proc { |context| context.example }
-
         config.around(:each) do |ex|
-          example = fetch_current_example.call(self)
+          ex.run_with_retry
+        end
+      end
+    end
+
+    attr_reader :context, :ex
 
-          retry_count = [
-            (
-              ENV['RSPEC_RETRY_RETRY_COUNT'] ||
+    def initialize(ex, opts = {})
+      @ex = ex
+      @ex.metadata.merge!(opts)
+      current_example.attempts ||= 0
+    end
+
+    # context.example is deprecated, but RSpec.current_example is not
+    # available until RSpec 3.0.
+    def current_example
+      RSpec.respond_to?(:current_example) ?
+        RSpec.current_example : @ex.example
+    end
+
+    def retry_count
+      [
+          (
+          ENV['RSPEC_RETRY_RETRY_COUNT'] ||
               ex.metadata[:retry] ||
               RSpec.configuration.default_retry_count
-            ).to_i,
-            1
-          ].max
-
-          sleep_interval = ex.metadata[:retry_wait] || RSpec.configuration.default_sleep_interval
-          exceptions_to_retry = ex.metadata[:exceptions_to_retry] || RSpec.configuration.exceptions_to_retry
-
-          clear_lets = ex.metadata[:clear_lets_on_failure]
-          clear_lets = RSpec.configuration.clear_lets_on_failure if clear_lets.nil?
-
-          retry_count.times do |i|
-            if RSpec.configuration.verbose_retry?
-              if i > 0
-                message = "RSpec::Retry: #{RSpec::Retry.ordinalize(i + 1)} try #{example.location}"
-                message = "\n" + message if i == 1
-                RSpec.configuration.reporter.message(message)
-              end
-            end
-            example.clear_exception
-            ex.run
-
-            break if example.exception.nil?
-
-            if exceptions_to_retry.any?
-              break unless exceptions_to_retry.any? do |exception_klass|
-                example.exception.is_a?(exception_klass)
-              end
-            end
-
-            if RSpec.configuration.verbose_retry? && RSpec.configuration.display_try_failure_messages?
-              if i != (retry_count-1)
-                try_message = "#{RSpec::Retry.ordinalize(i + 1)} Try error in #{example.location}:\n #{example.exception.to_s} \n"
-                RSpec.configuration.reporter.message(try_message)
-              end
-            end
-
-            self.clear_lets if clear_lets
-            sleep sleep_interval if sleep_interval.to_i > 0
+          ).to_i,
+          1
+      ].max
+    end
+
+    def attempts
+      current_example.attempts ||= 0
```

Comment:
The reason it's here is because it's possible for the `Retry` class to be instantiated and run multiple times. To avoid attempting too many times, the attempt count cannot be stored as a local variable in this class. I tried storing it in the example metadata but had trouble getting that to work.